### PR TITLE
Fixes issue when abandoning a multiple modal workflow

### PR DESCRIPTION
**Pull request by morganzellers**
## Checklist:
- [x] Is the linter reporting 0 errors?
- [ ] Did you comply with our [styleguide](https://github.com/wwt/Workflow/blob/main/STYLEGUIDE.md)?
- [x] Is there [adequate test coverage](https://github.com/wwt/Workflow/blob/main/CONTRIBUTING.md#test-etiquette) for your new code?
- [x] Does the CI pipeline pass?
- [x] Is the [static analysis tool](https://sonarcloud.io/dashboard?id=wwt_Workflow) reporting 0 issues?
- [ ] Did you [update the documentation](https://github.com/wwt/Workflow/blob/main/CONTRIBUTING.md#documentation)?
- [ ] Did you [update the sample app](https://github.com/wwt/Workflow/blob/main/CONTRIBUTING.md#sample-app)?
- [ ] Do we need to [increment the minor/major version](https://github.com/wwt/Workflow/blob/main/CONTRIBUTING.md#versioning)?
- [ ] Did you [change the public API](https://github.com/wwt/Workflow/blob/main/CONTRIBUTING.md#public-api)?
- [ ] Have you done everything you can to make sure that errors that can occur are compile-time errors, and if they have to be runtime do you have adequate tests and documentation around those runtime errors? [For more details](https://github.com/wwt/Workflow/blob/main/CONTRIBUTING.md#errors).

----

### If Applicable:
- [ ] Did you test when the first item is skipped?
- [ ] Did you test when the last item is skipped?
- [ ] Did you test when middle items are skipped?
- [ ] Did you test when incorrect data is passed forward?
- [ ] Did you test proceeding backwards?

----

### If Public API Has Changed:
- [ ] Did you deprecate (rather than remove) any old methods/variables/etc? [Our philosophy for deprecation](https://github.com/wwt/Workflow/blob/main/CONTRIBUTING.md#deprecation).
- [ ] Have you done the best that you can to make sure that the compiler guides people to changing to the new API? (Example: the renamed attribute)
- [ ] If necessary, have you tested the upgrade path for at least N-1 versions? For example, if data persists between v1 and v2 then that upgrade should be tested and as easy as we can make it.

### CODE REVIEW

**Comment by morganzellers** [comment]
Updated abandon function to be less specific around conditions for modal dismissal. - MZ BL

File: WorkflowExample/WorkflowUIKitTests/UIKitConsumerAbandonTests.swift [hunk 1]
```diff
@@ -219,6 +219,33 @@ class UIKitConsumerAbandonTests: XCTestCase {
...
+        XCTAssertUIViewControllerDisplayed(ofType: FR1.self)
+        XCTAssertNotNil(UIApplication.topViewController())
+        (UIApplication.topViewController() as? FR1)?.proceedInWorkflow(nil)
+        XCTAssertUIViewControllerDisplayed(ofType: FR2.self)
+        XCTAssertNil(UIApplication.topViewController()?.navigationController)
+        (UIApplication.topViewController() as? FR2)?.proceedInWorkflow(nil)
+        XCTAssertUIViewControllerDisplayed(ofType: FR3.self)
+        (UIApplication.topViewController() as? FR3)?.proceedInWorkflow(nil)
+        XCTAssertUIViewControllerDisplayed(ofType: FR4.self)
+        (UIApplication.topViewController() as? FR4)?.abandonWorkflow()
```
Comment (Richard-Gist) [inline WorkflowExample/WorkflowUIKitTests/UIKitConsumerAbandonTests.swift:246]: This is calling `abandonWorkflow()` on the `FlowRepresentable` but the issue calls out that it is about calling `abandon()` on the `Workflow`.  They ultimately end up in the same spot, but it's not exactly following the same code path as the bug.  I think this test is important to test what happens when a FlowRepresentable abandons the workflow in a chain of modals, but we need another test for when the workflow is abandoned.

File: Sources/WorkflowUIKit/UIKitPresenter.swift [hunk 2]
```diff
@@ -84,8 +84,8 @@ open class UIKitPresenter: OrchestrationResponder {
                     nav.popViewController(animated: animated)
                 }
             }
-        } else {
-            first.dismiss(animated: animated, completion: onFinish)
+        } else if let presenting = first.presentingViewController {
```
Comment (Richard-Gist) [inline Sources/WorkflowUIKit/UIKitPresenter.swift:87]: It looks like this has an unhandled situation for when there is no navigation controller and the view does not have a presentingViewController.   Is that a scenario that users could get into?  I think it's possible, but let me know if that has already been explored and deemed impossible.

**Review by Richard-Gist — CHANGES_REQUESTED** [review]

File: Sources/WorkflowUIKit/UIKitPresenter.swift (hunk 2 above)
Comment (morganzellers) [reply Sources/WorkflowUIKit/UIKitPresenter.swift:87]: Re-added the last `else` condition in. @brianlombardo and I had thought the `else if` we added would cover all scenarios, but it looks like there are tests still covering the last `else`, so I added back.

File: WorkflowExample/WorkflowUIKitTests/UIKitConsumerAbandonTests.swift (hunk 1 above)
Comment (morganzellers) [reply WorkflowExample/WorkflowUIKitTests/UIKitConsumerAbandonTests.swift:244]: Updated this test based on other requested fixes and added a new test for this requested fix

File: Sources/WorkflowUIKit/UIKitPresenter.swift [hunk 3]
```diff
@@ -86,6 +86,8 @@ open class UIKitPresenter: OrchestrationResponder {
             }
         } else if let presenting = first.presentingViewController {
             presenting.dismiss(animated: animated, completion: onFinish)
```
Comment (Richard-Gist) [inline Sources/WorkflowUIKit/UIKitPresenter.swift:88]: I could be misremembering how dismiss works, but doesn't it dismiss the view you call it on?  But this seems to be dismissing up to the view we are calling it on.  That makes me wonder if there are some exploratory scenarios we should look at (not necessarily unit test) around chaining workflows in different scenarios.

Or you know, if y'all already did all of that, then we're good and you can tell me to stop.

File: Sources/WorkflowUIKit/UIKitPresenter.swift (hunk 2 above)
Comment (Richard-Gist) [reply Sources/WorkflowUIKit/UIKitPresenter.swift:87]: There are tests covering them, but are they asserting on it?  I think tests that cover but not assert could be looked at for new tests that do assert.  Y'all have passed the pipeline with and without the `else` so I think we will still need to add tests around this.

**Review by Richard-Gist — CHANGES_REQUESTED** [review]

File: Sources/WorkflowUIKit/UIKitPresenter.swift (hunk 3 above)
Comment (wiemerm) [reply Sources/WorkflowUIKit/UIKitPresenter.swift:88]: We should check to see if having `ViewController` present another viewcontroller (I'll call this `ViewControllerB`) which in turn launches a workflow. Does the dismiss send us back to `ViewControllerB` or back to `ViewController` and which is the appropriate/expected behavior?

Comment (Richard-Gist) [reply Sources/WorkflowUIKit/UIKitPresenter.swift:88]: Sending back to `ViewControllerB` is what I think the correct behavior should be.  I've abandoned my Workflow so all things related to my workflow I closed, but not the thing that launched the workflow.

Comment (morganzellers) [reply Sources/WorkflowUIKit/UIKitPresenter.swift:88]: `dismiss` has separate functionalities depending on the calling view controller.  We are forcing it to always use the behavior where the calling view controller relies on the presenting view controller to dismiss.

File: Sources/WorkflowUIKit/UIKitPresenter.swift (hunk 2 above)
Comment (morganzellers) [reply Sources/WorkflowUIKit/UIKitPresenter.swift:87]: Xcode was reporting coverage of the else, but we were unable to hit breakpoints, print(), or fatalError() functions inside of the else. As such, we removed the untestable line.

**Review by Richard-Gist — APPROVED** [review]
Thank you for all the hard work!

**Pull request merged**